### ISSUE TITLE
False negative for CKV_AWS_24 (unrestricted SSH ingress)

### ISSUE THREAD

**Issue by acdha** [bridgecrewio/checkov#264]
**Describe the bug**

The CKV_AWS_24 rule appears to have stopped working at some point.

**To Reproduce**
Run checkov — e.g. `docker run -it --rm -v $(pwd):/code bridgecrew/checkov -d /code` — using this Terraform file:

<details>

```hcl
resource "aws_security_group" "inline_rules" {
  description = "SG with inline rules"
  ingress = [
    {
      cidr_blocks      = ["0.0.0.0/0"]
      description      = "Wide Open"
      from_port        = 0
      ipv6_cidr_blocks = []
      prefix_list_ids  = []
      security_groups  = []
      protocol         = "-1"
      self             = false
      to_port          = 65535
    }
  ]
}

resource "aws_security_group" "separate_rules" {
  description = "SG with separate rules"
}

resource "aws_security_group_rule" "bad" {
  description       = "Allow inbound SSH from the world!"
  type              = "ingress"
  security_group_id = aws_security_group.separate_rules.id
  protocol          = "tcp"
  from_port         = 22
  to_port           = 22
  cidr_blocks       = ["0.0.0.0/0"]
}

resource "aws_security_group_rule" "worse" {
  description       = "Allow everything!"
  type              = "ingress"
  security_group_id = aws_security_group.separate_rules.id
  protocol          = "-1"
  from_port         = 0
  to_port           = 65535
  cidr_blocks       = ["0.0.0.0/0"]
}
```

</details>

**Expected behavior**

The security groups should be flagged

**Screenshots**

**Desktop (please complete the following information):**
 - OS: MacOS using Docker.app
 - Checkov Version: version: 1.0.290 

```
Using default tag: latest
latest: Pulling from bridgecrew/checkov
Digest: sha256:8ad4123715f6c75df75810aecea0b861e02dbe37dee823f4d2514147d7dd64d7
Status: Image is up to date for bridgecrew/checkov:latest
docker.io/bridgecrew/checkov:latest
```

**Additional context**

<details>

```
docker run -it --rm -v (pwd):/code bridgecrew/checkov -d /code

       _               _              
   ___| |__   ___  ___| | _______   __
  / __| '_ \ / _ \/ __| |/ / _ \ \ / /
 | (__| | | |  __/ (__|   < (_) \ V / 
  \___|_| |_|\___|\___|_|\_\___/ \_/  
                                      
version: 1.0.290 

terraform scan results:

Passed checks: 8, Failed checks: 0, Skipped checks: 0

Check: CKV_AWS_25: "Ensure no security groups allow ingress from 0.0.0.0:0 to port 3389"
	PASSED for resource: aws_security_group.inline_rules
	File: /security-groups.tf:1-16


Check: CKV_AWS_23: "Ensure every security groups rule has a description"
	PASSED for resource: aws_security_group.inline_rules
	File: /security-groups.tf:1-16


Check: CKV_AWS_24: "Ensure no security groups allow ingress from 0.0.0.0:0 to port 22"
	PASSED for resource: aws_security_group.inline_rules
	File: /security-groups.tf:1-16


Check: CKV_AWS_25: "Ensure no security groups allow ingress from 0.0.0.0:0 to port 3389"
	PASSED for resource: aws_security_group.separate_rules
	File: /security-groups.tf:18-20


Check: CKV_AWS_23: "Ensure every security groups rule has a description"
	PASSED for resource: aws_security_group.separate_rules
	File: /security-groups.tf:18-20


Check: CKV_AWS_24: "Ensure no security groups allow ingress from 0.0.0.0:0 to port 22"
	PASSED for resource: aws_security_group.separate_rules
	File: /security-groups.tf:18-20


Check: CKV_AWS_23: "Ensure every security groups rule has a description"
	PASSED for resource: aws_security_group_rule.bad
	File: /security-groups.tf:22-30


Check: CKV_AWS_23: "Ensure every security groups rule has a description"
	PASSED for resource: aws_security_group_rule.worse
	File: /security-groups.tf:33-41
```
</details>

**Comment by acdha** [comment]
A quick pass through pdb revealed 3 problems:

1. The type check on https://github.com/bridgecrewio/checkov/blob/5658c3cb762ef6a0bf51903466f94e50754de019/checkov/terraform/checks/resource/aws/SecurityGroupUnrestrictedIngress22.py#L26 is expecting a dictionary but receives a list of dictionaries. It proceeds further if you do something hacky like `exec("rule = rule[0]")`.
2. The type check fails silently — it should probably either just assume a dictionary and fail with an exception if given something else or have an else path which raises a `TypeError`.
3. The port check similarly hits a changed type signature causing a silent failure: it's comparing an integer to a list containing an integer.
4. The port check only catches exact matches. It should be doing a numeric test to catch cases like from_port = 0 / to_port = 65535

**Comment by nimrodkor** [comment]
Hi @acdha , Can you please provide an example terraform code of something that fails currently?

**Comment by acdha** [comment]
The example in the description has two forms of the error, both of which are currently passed by Checkov. I noticed this when Security Hub flagged something which Checkov had passed on a project which didn’t also use tfsec (which did flag it).

**Comment by nimrodkor** [comment]
@acdha Thanks for reporting!
I've gone ahead and refactored the SG rule checks - they didn't support the format of a list of ingress rules.

Thanks for using checkov and reporting this issue!
